File: v2/rp_runtime/src/lib/semantic-qa-envelope.mjs
import { parseJsonObject } from './inference-utils.mjs';

export const SEMANTIC_QA_RESULT_SCHEMA = 'hg_semantic_qa_result_v1';

export const VALID_EVALUATION_TARGET_ROLES = new Set(['director', 'narrator', 'character']);
export const VALID_OVERALL_RESULTS = new Set(['pass', 'reject_soft', 'reject_hard']);
export const VALID_SEVERITIES = new Set(['hard', 'soft']);

export const CITATION_STATUS = {
  VALID: 'valid',
  UNKNOWN_REF: 'unknown_ref',
  MISSING_CITATION: 'missing_citation',
  ADVISORY_AUTHORITY_CLASS: 'advisory_authority_class',
  DERIVED_AUTHORITY_CLASS: 'derived_authority_class',
};

const SEMANTIC_QA_AUTHORITY_CLASSES = new Set(['authoritative', 'derived', 'advisory']);

export function normalizeAuthorityClass(value) {
  const text = String(value ?? '').trim().toLowerCase();
  return SEMANTIC_QA_AUTHORITY_CLASSES.has(text) ? text : null;
}

export function validateAuthorityReference(ref) {
  const errors = [];
  if (!ref || typeof ref !== 'object') {
    return ['authority reference must be an object'];
  }
  const refId = String(ref.ref_id ?? ref.authority_ref_id ?? '').trim();
  if (!refId) errors.push('ref_id is required');
  if (!String(ref.kind ?? '').trim()) errors.push('kind is required');
  if (!normalizeAuthorityClass(ref.authority_class)) {
    errors.push('authority_class must be authoritative, derived, or advisory');
  }
  if (!String(ref.label ?? '').trim()) errors.push('label is required');
  if (ref.text == null || String(ref.text).trim() === '') errors.push('text is required');
  return errors;
}

export function validateAuthorityReferences(refs) {
  if (!Array.isArray(refs)) {
    return { normalized: [], errors: ['authority_references must be a list'] };
  }
  const normalized = [];
  const errors = [];
  const seen = new Set();
  for (let index = 0; index < refs.length; index += 1) {
    const itemErrors = validateAuthorityReference(refs[index]);
    if (itemErrors.length) {
      errors.push(...itemErrors.map((msg) => `authority_references[${index}]: ${msg}`));
      continue;
    }
    const refId = String(refs[index].ref_id ?? refs[index].authority_ref_id ?? '').trim();
    if (seen.has(refId)) {
      errors.push(`authority_references[${index}]: duplicate ref_id ${refId}`);
      continue;
    }
    seen.add(refId);
    normalized.push({
      ref_id: refId,
      kind: String(refs[index].kind ?? '').trim(),
      authority_class: normalizeAuthorityClass(refs[index].authority_class),
      label: String(refs[index].label ?? '').trim(),
      text: String(refs[index].text ?? ''),
      ...(refs[index].provenance && typeof refs[index].provenance === 'object'
        ? { provenance: { ...refs[index].provenance } }
        : {}),
    });
  }
  return { normalized, errors };
}

export function authorityReferenceIndex(refs) {
  const index = new Map();
  for (const ref of refs ?? []) {
    const refId = String(ref?.ref_id ?? '').trim();
    if (refId) index.set(refId, ref);
  }
  return index;
}

export function validateFindingCitations(findings, authorityRefs) {
  const refIndex = authorityReferenceIndex(authorityRefs);
  const validations = [];
  for (let findingIndex = 0; findingIndex < (findings ?? []).length; findingIndex += 1) {
    const finding = findings[findingIndex];
    const severity = String(finding?.severity ?? '').trim();
    const citation = finding?.authoritative_citation;
    const refId = citation && typeof citation === 'object'
      ? String(citation.ref_id ?? citation.authority_ref_id ?? '').trim()
      : '';

    if (severity === 'hard' && !refId) {
      validations.push({
        finding_index: findingIndex,
        ref_id: null,
        status: CITATION_STATUS.MISSING_CITATION,
        resolved_authority_class: null,
      });
      continue;
    }
    if (!refId) continue;

    const resolved = refIndex.get(refId);
    if (!resolved) {
      validations.push({
        finding_index: findingIndex,
        ref_id: refId,
        status: CITATION_STATUS.UNKNOWN_REF,
        resolved_authority_class: null,
      });
      continue;
    }

    const resolvedClass = normalizeAuthorityClass(resolved.authority_class);
    let status = CITATION_STATUS.VALID;
    if (resolvedClass === 'advisory') status = CITATION_STATUS.ADVISORY_AUTHORITY_CLASS;
    else if (resolvedClass === 'derived') status = CITATION_STATUS.DERIVED_AUTHORITY_CLASS;

    validations.push({
      finding_index: findingIndex,
      ref_id: refId,
      status,
      resolved_authority_class: resolvedClass,
    });
  }
  return validations;
}

function normalizeFinding(raw, findingIndex) {
  const warnings = [];
  if (!raw || typeof raw !== 'object') {
    warnings.push(`findings[${findingIndex}] is not an object`);
    return { finding: null, warnings };
  }
  const dimension = String(raw.dimension ?? '').trim();
  if (!dimension) {
    warnings.push(`findings[${findingIndex}].dimension is required`);
    return { finding: null, warnings };
  }
  const severity = String(raw.severity ?? '').trim();
  if (!VALID_SEVERITIES.has(severity)) {
    warnings.push(`findings[${findingIndex}].severity must be hard or soft`);
    return { finding: null, warnings };
  }

  let normalizedCitation = null;
  if (raw.authoritative_citation && typeof raw.authoritative_citation === 'object') {
    const refId = String(
      raw.authoritative_citation.ref_id
      ?? raw.authoritative_citation.authority_ref_id
      ?? '',
    ).trim();
    if (refId) normalizedCitation = { ref_id: refId };
  } else if (severity === 'hard') {
    warnings.push(
      `findings[${findingIndex}] hard severity requires authoritative_citation.ref_id`,
    );
  }

  return {
    finding: {
      dimension,
      severity,
      finding: String(raw.finding ?? ''),
      rationale: String(raw.rationale ?? ''),
      candidate_evidence: raw.candidate_evidence ?? null,
      authoritative_citation: normalizedCitation,
    },
    warnings,
  };
}

export function parseSemanticQaResult(raw, authorityReferences = [], options = {}) {
  const {
    expectedEvaluationPassId = null,
    expectedEvaluationTargetRole = null,
  } = options;

  let parsed;
  try {
    parsed = typeof raw === 'string' ? parseJsonObject(raw) : raw;
  } catch (error) {
    return {
      ok: false,
      error: String(error),
      result: null,
      citationValidations: [],
      parseWarnings: [],
    };
  }
  if (!parsed || typeof parsed !== 'object') {
    return {
      ok: false,
      error: 'evaluator output not an object',
      result: null,
      citationValidations: [],
      parseWarnings: [],
    };
  }

  const parseWarnings = [];
  if (String(parsed.schema ?? '').trim() !== SEMANTIC_QA_RESULT_SCHEMA) {
    return {
      ok: false,
      error: `unsupported schema ${String(parsed.schema ?? '<missing>')}`,
      result: null,
      citationValidations: [],
      parseWarnings: [],
    };
  }

  const evaluationTargetRole = String(parsed.evaluation_target_role ?? '').trim();
  if (!VALID_EVALUATION_TARGET_ROLES.has(evaluationTargetRole)) {
    return {
      ok: false,
      error: 'evaluation_target_role must be director, narrator, or character',
      result: null,
      citationValidations: [],
      parseWarnings: [],
    };
  }
  if (expectedEvaluationTargetRole && evaluationTargetRole !== expectedEvaluationTargetRole) {
    parseWarnings.push('evaluation_target_role does not match expected evaluation target');
  }

  const evaluationPassId = String(parsed.evaluation_pass_id ?? '').trim();
  if (!evaluationPassId) {
    return {
      ok: false,
      error: 'evaluation_pass_id is required',
      result: null,
      citationValidations: [],
      parseWarnings: [],
    };
  }
  if (expectedEvaluationPassId && evaluationPassId !== expectedEvaluationPassId) {
    parseWarnings.push('evaluation_pass_id does not match expected correlation id');
  }

  const overallResult = String(parsed.overall_result ?? '').trim();
  if (!VALID_OVERALL_RESULTS.has(overallResult)) {
    return {
      ok: false,
      error: 'overall_result must be pass, reject_soft, or reject_hard',
      result: null,
      citationValidations: [],
      parseWarnings: [],
    };
  }

  const { normalized: refs, errors: refErrors } = validateAuthorityReferences(
    authorityReferences ?? [],
  );
  parseWarnings.push(...refErrors);

  const findingsRaw = Array.isArray(parsed.findings) ? parsed.findings : null;
  if (findingsRaw === null) {
    return {
      ok: false,
      error: 'findings must be an array',
      result: null,
      citationValidations: [],
      parseWarnings: [],
    };
  }

  const findings = [];
  for (let index = 0; index < findingsRaw.length; index += 1) {
    const { finding, warnings } = normalizeFinding(findingsRaw[index], index);
    parseWarnings.push(...warnings);
    if (finding) findings.push(finding);
  }

  const citationValidations = validateFindingCitations(findings, refs);
  const evaluatorSummary = parsed.evaluator_summary;
  if (evaluatorSummary != null && typeof evaluatorSummary !== 'string') {
    parseWarnings.push('evaluator_summary must be a string when present');
  }

  let residualSoftConcerns = [];
  if (Array.isArray(parsed.residual_soft_concerns)) {
    residualSoftConcerns = parsed.residual_soft_concerns.map((item) => String(item));
  } else if (parsed.residual_soft_concerns != null) {
    parseWarnings.push('residual_soft_concerns must be an array when present');
  }

  return {
    ok: true,
    error: null,
    result: {
      schema: SEMANTIC_QA_RESULT_SCHEMA,
      evaluation_target_role: evaluationTargetRole,
      evaluation_pass_id: evaluationPassId,
      overall_result: overallResult,
      findings,
      evaluator_summary: evaluatorSummary ? String(evaluatorSummary) : null,
      residual_soft_concerns: residualSoftConcerns,
    },
    citationValidations,
    parseWarnings,
  };
}

File: v2/domain/modules/semantic_qa_envelope.py
"""Shared semantic-QA result envelope parser (#25)."""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any

from .authority_reference import (
    validate_authority_references,
    validate_finding_citations,
)

SEMANTIC_QA_RESULT_SCHEMA = "hg_semantic_qa_result_v1"
VALID_EVALUATION_TARGET_ROLES = frozenset({"director", "narrator", "character"})
VALID_OVERALL_RESULTS = frozenset({"pass", "reject_soft", "reject_hard"})
VALID_SEVERITIES = frozenset({"hard", "soft"})


@dataclass(frozen=True)
class SemanticQaParseResult:
    ok: bool
    error: str | None
    result: dict[str, Any] | None
    citation_validations: tuple[dict[str, Any], ...] = ()
    parse_warnings: tuple[str, ...] = ()


def _load_json_object(raw: Any) -> dict[str, Any]:
    if isinstance(raw, dict):
        return raw
    if isinstance(raw, str):
        parsed = json.loads(raw)
        if not isinstance(parsed, dict):
            raise ValueError("evaluator output not an object")
        return parsed
    raise ValueError("evaluator output must be JSON object or string")


def _normalize_finding(raw: Any, *, finding_index: int) -> tuple[dict[str, Any] | None, list[str]]:
    warnings: list[str] = []
    if not isinstance(raw, dict):
        warnings.append(f"findings[{finding_index}] is not an object")
        return None, warnings

    dimension = str(raw.get("dimension") or "").strip()
    if not dimension:
        warnings.append(f"findings[{finding_index}].dimension is required")
        return None, warnings

    severity = str(raw.get("severity") or "").strip()
    if severity not in VALID_SEVERITIES:
        warnings.append(
            f"findings[{finding_index}].severity must be hard or soft",
        )
        return None, warnings

    citation = raw.get("authoritative_citation")
    normalized_citation = None
    if isinstance(citation, dict):
        ref_id = str(
            citation.get("ref_id") or citation.get("authority_ref_id") or "",
        ).strip()
        if ref_id:
            normalized_citation = {"ref_id": ref_id}
    elif severity == "hard":
        warnings.append(
            f"findings[{finding_index}] hard severity requires authoritative_citation.ref_id",
        )

    finding = {
        "dimension": dimension,
        "severity": severity,
        "finding": str(raw.get("finding") or ""),
        "rationale": str(raw.get("rationale") or ""),
        "candidate_evidence": raw.get("candidate_evidence"),
        "authoritative_citation": normalized_citation,
    }
    return finding, warnings


def parse_semantic_qa_result(
    raw: Any,
    authority_references: list[dict[str, Any]] | None = None,
    *,
    expected_evaluation_pass_id: str | None = None,
    expected_evaluation_target_role: str | None = None,
) -> SemanticQaParseResult:
    """Parse hg_semantic_qa_result_v1 without applying role semantic policy."""
    try:
        parsed = _load_json_object(raw)
    except (json.JSONDecodeError, ValueError, TypeError) as error:
        return SemanticQaParseResult(ok=False, error=str(error), result=None)

    warnings: list[str] = []

    schema = str(parsed.get("schema") or "").strip()
    if schema != SEMANTIC_QA_RESULT_SCHEMA:
        return SemanticQaParseResult(
            ok=False,
            error=f"unsupported schema {schema or '<missing>'}",
            result=None,
        )

    evaluation_target_role = str(parsed.get("evaluation_target_role") or "").strip()
    if evaluation_target_role not in VALID_EVALUATION_TARGET_ROLES:
        return SemanticQaParseResult(
            ok=False,
            error="evaluation_target_role must be director, narrator, or character",
            result=None,
        )
    if (
        expected_evaluation_target_role
        and evaluation_target_role != expected_evaluation_target_role
    ):
        warnings.append(
            "evaluation_target_role does not match expected evaluation target",
        )

    evaluation_pass_id = str(parsed.get("evaluation_pass_id") or "").strip()
    if not evaluation_pass_id:
        return SemanticQaParseResult(
            ok=False,
            error="evaluation_pass_id is required",
            result=None,
        )
    if expected_evaluation_pass_id and evaluation_pass_id != expected_evaluation_pass_id:
        warnings.append("evaluation_pass_id does not match expected correlation id")

    overall_result = str(parsed.get("overall_result") or "").strip()
    if overall_result not in VALID_OVERALL_RESULTS:
        return SemanticQaParseResult(
            ok=False,
            error="overall_result must be pass, reject_soft, or reject_hard",
            result=None,
        )

    refs_input = authority_references if authority_references is not None else []
    normalized_refs, ref_errors = validate_authority_references(refs_input)
    if ref_errors:
        warnings.extend(ref_errors)

    findings_raw = parsed.get("findings")
    if findings_raw is None:
        findings_raw = []
    if not isinstance(findings_raw, list):
        return SemanticQaParseResult(
            ok=False,
            error="findings must be an array",
            result=None,
        )

    findings: list[dict[str, Any]] = []
    for index, item in enumerate(findings_raw):
        finding, finding_warnings = _normalize_finding(item, finding_index=index)
        warnings.extend(finding_warnings)
        if finding is not None:
            findings.append(finding)

    citation_validations = validate_finding_citations(findings, normalized_refs)

    evaluator_summary = parsed.get("evaluator_summary")
    if evaluator_summary is not None and not isinstance(evaluator_summary, str):
        warnings.append("evaluator_summary must be a string when present")

    residual = parsed.get("residual_soft_concerns")
    if residual is None:
        residual_list: list[Any] = []
    elif isinstance(residual, list):
        residual_list = [str(item) for item in residual]
    else:
        warnings.append("residual_soft_concerns must be an array when present")
        residual_list = []

    result = {
        "schema": SEMANTIC_QA_RESULT_SCHEMA,
        "evaluation_target_role": evaluation_target_role,
        "evaluation_pass_id": evaluation_pass_id,
        "overall_result": overall_result,
        "findings": findings,
        "evaluator_summary": str(evaluator_summary) if evaluator_summary else None,
        "residual_soft_concerns": residual_list,
    }

    return SemanticQaParseResult(
        ok=True,
        error=None,
        result=result,
        citation_validations=tuple(citation_validations),
        parse_warnings=tuple(warnings),
    )
